### rozdelenie práce 50/50

In [104]:
import pandas as pd

## Popis parametrov senzorov

In [105]:
sensor = pd.read_csv('sensor_variable_range.csv',sep='\t')
sensor

Variable         Unit  \
0                    SpO₂            %   
1                      HR          bpm   
2                      PI            %   
3                      RR  breaths/min   
4                   EtCO₂         mmHg   
5                    FiO₂            %   
6                     PRV           ms   
7                      BP         mmHg   
8        Skin Temperature           °C   
9   Motion/Activity index            -   
10                    PVI            %   
11               Hb level         g/dL   
12                     SV      mL/beat   
13                     CO        L/min   
14       Blood Flow Index            -   
15  PPG waveform features            -   
16   Signal Quality Index            -   
17     Respiratory effort            -   
18    O₂ extraction ratio            -   
19                    SNR           dB   

                                     Description        Value Range  
0                   Peripheral oxygen saturation            95–100%  
1                 Heart rate from pulse oximeter         60–100 bpm  
2                                Perfusion index            0.2–20%  
3                               Respiratory rate  12–20 breaths/min  
4                       End-tidal carbon dioxide         35–45 mmHg  
5                    Fraction of inspired oxygen            21–100%  
6                         Pulse rate variability          20–200 ms  
7       Blood pressure (systolic diastolic mean)  90/60–120/80 mmHg  
8                         Peripheral temperature           33–38 °C  
9                     Sensor artifact correction                NaN  
10                       Pleth Variability Index             10–20%  
11                      Hemoglobin concentration         12–18 g/dL  
12                                 Stroke volume     60–100 mL/beat  
13                                Cardiac output          4–8 L/min  
14  Optical/near-infrared blood flow measurement     Relative units  
15        Photoplethysmography signal parameters                NaN  
16             Reliability indicator of readings              0–100  
17        Chest impedance or belt-derived effort                NaN  
18           Calculated oxygen extraction metric            0.2–0.3  
19             Signal-to-noise ratio from sensor           20–40 dB

# 1.1 Základný opis dát spolu s ich charakteristikami

## Načítanie dát do premenných

In [106]:
df_observation = pd.read_csv('datasets/observation.csv',sep='\t')
df_patient = pd.read_csv('datasets/patient.csv',sep='\t')
df_station = pd.read_csv('datasets/station.csv',sep='\t')

#### 1. Prieskum dát - základné štatistiky a informácie o dátach
1. `info()`
2. `isnul().sum()`
3. `head()`
4. `nunique()` - len unique hodnot
5. `describe()` - štatistiky

#### 2. Analýza atribútov
1. Typy atribútov - `select_dtypes()`
    - `select_dtypes().columns`
    - `select_dtypes().nunique()`

#### 3. Vysledok analýzy - poznámky o dátach, návrhy na úpravy

## Dataset Observation

### 1. Prieskum dát

In [107]:
df_observation.shape

(12047, 23)

In [108]:
df_observation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12047 entries, 0 to 12046
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   SpO₂                   12047 non-null  float64
 1   HR                     12047 non-null  float64
 2   PI                     12047 non-null  float64
 3   RR                     12047 non-null  float64
 4   EtCO₂                  12047 non-null  float64
 5   FiO₂                   12047 non-null  float64
 6   PRV                    12047 non-null  float64
 7   BP                     12047 non-null  float64
 8   Skin Temperature       12047 non-null  float64
 9   Motion/Activity index  12047 non-null  float64
 10  PVI                    12047 non-null  float64
 11  Hb level               12047 non-null  float64
 12  SV                     12047 non-null  float64
 13  CO                     12047 non-null  float64
 14  Blood Flow Index       12047 non-null  float64
 15  PP

In [109]:
df_observation.isnull().sum()

SpO₂                     0
HR                       0
PI                       0
RR                       0
EtCO₂                    0
FiO₂                     0
PRV                      0
BP                       0
Skin Temperature         0
Motion/Activity index    0
PVI                      0
Hb level                 0
SV                       0
CO                       0
Blood Flow Index         0
PPG waveform features    0
Signal Quality Index     0
Respiratory effort       0
O₂ extraction ratio      0
SNR                      0
oximetry                 0
latitude                 0
longitude                0
dtype: int64

In [110]:
df_observation.head()

SpO₂         HR         PI         RR      EtCO₂       FiO₂  \
0  97.620193  73.047425  14.796690  14.368778  41.147562  44.998343   
1  97.828560  91.254424  15.226811  17.746857  38.822901  67.082312   
2  96.771755  80.281644   9.746972  18.608674  37.259533  30.457594   
3  97.060000  78.424910  10.731992  18.942426  37.970214  51.181183   
4  96.852482  77.855986   9.266719  14.754602  43.034371  49.059859   

          PRV          BP  Skin Temperature  Motion/Activity index  ...  \
0  101.744512  108.175478         35.284983              11.207029  ...   
1  136.925626  106.173834         35.174372               7.212819  ...   
2  117.224898  106.045090         36.168580              13.639960  ...   
3  105.148959  107.950747         36.229369               4.946382  ...   
4  140.404608  104.395081         35.566732              13.013934  ...   

         CO  Blood Flow Index  PPG waveform features  Signal Quality Index  \
0  4.024120         57.393854              31.011871             61.170762   
1  4.745712         27.416950              70.006621             36.441271   
2  4.094455         58.161639              61.632568             59.933116   
3  4.064535         44.622856              64.227237             32.365137   
4  4.060419         55.895316              71.565659             65.842228   

   Respiratory effort  O₂ extraction ratio        SNR  oximetry  latitude  \
0           84.520164             0.292173  22.855412       1.0  53.69786   
1           39.225366             0.210614  33.204156       0.0  53.59782   
2           55.132717             0.238559  31.176268       1.0  10.73750   
3           69.857796             0.225224  30.328586       0.0  12.25282   
4           62.507491             0.214413  20.550401       1.0 -12.91667   

   longitude  
0   -2.68758  
1   34.33825  
2  122.96660  
3   79.41727  
4  -39.25000  

[5 rows x 23 columns]

In [111]:
df_observation.nunique()

SpO₂                     11869
HR                       11869
PI                       11869
RR                       11869
EtCO₂                    11869
FiO₂                     11869
PRV                      11869
BP                       11869
Skin Temperature         11869
Motion/Activity index    11869
PVI                      11869
Hb level                 11869
SV                       11869
CO                       11869
Blood Flow Index         11869
PPG waveform features    11869
Signal Quality Index     11869
Respiratory effort       11869
O₂ extraction ratio      11869
SNR                      11869
oximetry                     2
latitude                   558
longitude                  558
dtype: int64

In [112]:
df_observation.describe()

SpO₂            HR            PI            RR         EtCO₂  \
count  12047.000000  12047.000000  12047.000000  12047.000000  12047.000000   
mean      97.267672     80.697806      9.833910     16.127032     40.492596   
std        0.598524      5.155163      2.700888      1.341912      1.726204   
min       95.000000     60.000000      0.200000     12.000000     35.000000   
25%       96.866478     77.270972      8.026067     15.065575     39.148665   
50%       97.264309     80.698631      9.839743     15.910760     40.725808   
75%       97.668347     84.159349     11.637860     17.268908     41.860040   
max      100.000000    100.000000     20.000000     20.000000     45.000000   

               FiO₂           PRV            BP  Skin Temperature  \
count  12047.000000  12047.000000  12047.000000      12047.000000   
mean      56.430377    104.028280    104.556883         35.712090   
std        9.113699     21.815668      3.718477          0.606053   
min       21.000000     20.000000     90.000000         33.000000   
25%       50.319755     89.683443    102.051506         35.305665   
50%       56.215376    103.995530    104.548575         35.712252   
75%       62.411973    118.315006    107.076850         36.113696   
max      100.000000    200.000000    120.000000         38.000000   

       Motion/Activity index  ...            CO  Blood Flow Index  \
count           12047.000000  ...  12047.000000      12047.000000   
mean               10.696255  ...      4.161154         52.047837   
std                 3.115040  ...      0.204165         12.529413   
min                 0.476937  ...      4.000000          0.000000   
25%                 8.396510  ...      4.052809         43.541329   
50%                10.790450  ...      4.099662         52.152080   
75%                13.261020  ...      4.192040         60.604392   
max                19.452927  ...      8.000000        100.000000   

       PPG waveform features  Signal Quality Index  Respiratory effort  \
count           12047.000000          12047.000000        12047.000000   
mean               57.260950             49.055111           53.443099   
std                12.585968             13.431124           13.873025   
min                 0.000000              0.000000            0.000000   
25%                48.839388             39.961822           44.047538   
50%                57.212810             48.931583           53.517314   
75%                65.718705             58.244514           62.875562   
max               100.000000            100.000000          100.000000   

       O₂ extraction ratio           SNR      oximetry      latitude  \
count         12047.000000  12047.000000  12047.000000  12047.000000   
mean              0.249607     29.971706      0.595335     28.074873   
std               0.028951      5.761003      0.490847     23.392038   
min               0.200000     20.000000      0.000000    -54.800000   
25%               0.224547     24.967058      0.000000     13.566670   
50%               0.249234     30.044146      1.000000     34.864720   
75%               0.274752     34.936653      1.000000     45.780710   
max               0.300000     40.000000      1.000000     68.798330   

          longitude  
count  12047.000000  
mean      14.764536  
std       71.362268  
min     -123.026210  
25%      -45.947220  
50%       13.576610  
75%       72.851070  
max      171.253640  

[8 rows x 23 columns]

### 2. Analýza atribútov

In [113]:
len(df_observation.select_dtypes(include=['float64']).columns)

23

In [114]:
df_observation.select_dtypes(include=['float64']).columns

Index(['SpO₂', 'HR', 'PI', 'RR', 'EtCO₂', 'FiO₂', 'PRV', 'BP',
       'Skin Temperature', 'Motion/Activity index', 'PVI', 'Hb level', 'SV',
       'CO', 'Blood Flow Index', 'PPG waveform features',
       'Signal Quality Index', 'Respiratory effort', 'O₂ extraction ratio',
       'SNR', 'oximetry', 'latitude', 'longitude'],
      dtype='object')

In [115]:
df_observation.select_dtypes(include=['float64']).nunique()

SpO₂                     11869
HR                       11869
PI                       11869
RR                       11869
EtCO₂                    11869
FiO₂                     11869
PRV                      11869
BP                       11869
Skin Temperature         11869
Motion/Activity index    11869
PVI                      11869
Hb level                 11869
SV                       11869
CO                       11869
Blood Flow Index         11869
PPG waveform features    11869
Signal Quality Index     11869
Respiratory effort       11869
O₂ extraction ratio      11869
SNR                      11869
oximetry                     2
latitude                   558
longitude                  558
dtype: int64

### 3. Vysledok analýzy
- dataset pozostáva z 23 atribútov a 12047 záznamov
- všetky atribúty sú číselné typu `float64` - ide o kontinuálne merania z rôznych senzorov
- neobsahuje žiadne chýbajúce hodnoty



## Dataset Patient

### 1. Prieskum dát

In [116]:
df_patient.shape

(2069, 13)

In [117]:
df_patient.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2069 entries, 0 to 2068
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   mail              2069 non-null   object 
 1   birthdate         1138 non-null   object 
 2   registration      2069 non-null   object 
 3   blood_group       2069 non-null   object 
 4   company           2069 non-null   object 
 5   username          2069 non-null   object 
 6   user_id           2069 non-null   int64  
 7   ssn               2069 non-null   object 
 8   residence         0 non-null      float64
 9   name              2069 non-null   object 
 10  job               621 non-null    object 
 11  current_location  1966 non-null   object 
 12  station_ID        2069 non-null   int64  
dtypes: float64(1), int64(2), object(10)
memory usage: 210.3+ KB


In [118]:
df_patient.isnull().sum()

mail                   0
birthdate            931
registration           0
blood_group            0
company                0
username               0
user_id                0
ssn                    0
residence           2069
name                   0
job                 1448
current_location     103
station_ID             0
dtype: int64

In [119]:
df_patient.head()

mail   birthdate registration blood_group  \
0  alicevasconcelos@bol.com.br  2014-05-22  03 Jul 2022          B+   
1          carla33@hotmail.com         NaN   2019-02-19         AB-   
2            julie13@yahoo.com         NaN  18 Apr 2023          B+   
3    murakamishota@hotmail.com         NaN   2021-05-31          A+   
4            mauro58@gmail.com         NaN   2021-10-19          A-   

                  company      username  user_id               ssn  residence  \
0          Cavalcanti S/A        theo31      763       38671502902        NaN   
1              Buck-Brown      donald76      571       648-52-7110        NaN   
2            Acosta-Black    diazrhonda     1368       472-71-8574        NaN   
3                加藤農林合同会社   nanamiinoue     1011       156-44-9296        NaN   
4  Ferragni-Rapisardi SPA  gaetanoboito     1662  CMMSYV42C16A763E        NaN   

                  name     job  \
0        Yuri Mendonça     NaN   
1       Amanda Douglas     NaN   
2          Jordan Shea     NaN   
3                伊藤 修平  高等学校教員   
4  Sig. Atenulf Polani     NaN   

                                  current_location  station_ID  
0   (Decimal('75.5622165'), Decimal('140.611314'))         705  
1    (Decimal('56.355046'), Decimal('-31.498637'))         387  
2  (Decimal('77.9571885'), Decimal('-141.860092'))         593  
3      (Decimal('5.799277'), Decimal('58.160125'))         429  
4  (Decimal('78.8543295'), Decimal('-101.613684'))         388

In [120]:
df_patient.nunique()

mail                2062
birthdate           1119
registration        1875
blood_group            8
company             1885
username            2042
user_id             1315
ssn                 2069
residence              0
name                2018
job                  406
current_location    1966
station_ID           767
dtype: int64

In [121]:
df_patient.describe()

user_id  residence   station_ID
count  2069.000000        0.0  2069.000000
mean   1030.209763        NaN   409.358627
std     606.638857        NaN   244.337210
min       0.000000        NaN     0.000000
25%     506.000000        NaN   193.000000
50%     999.000000        NaN   406.000000
75%    1571.000000        NaN   625.000000
max    2065.000000        NaN   830.000000

In [122]:
df_patient.groupby('user_id').head()

mail   birthdate          registration  \
0     alicevasconcelos@bol.com.br  2014-05-22           03 Jul 2022   
1             carla33@hotmail.com         NaN            2019-02-19   
2               julie13@yahoo.com         NaN           18 Apr 2023   
3       murakamishota@hotmail.com         NaN            2021-05-31   
4               mauro58@gmail.com         NaN            2021-10-19   
...                           ...         ...                   ...   
2064     naokitakahashi@yahoo.com         NaN            2019-01-14   
2065      igordas-neves@gmail.com         NaN            2019/11/22   
2066   lucianoletta@fastwebnet.it  2022-03-09           11 Jun 2024   
2067           mary08@hotmail.com         NaN           02 Feb 2021   
2068          clayton74@yahoo.com         NaN  10/25/2020, 00:00:00   

     blood_group                 company            username  user_id  \
0             B+          Cavalcanti S/A              theo31      763   
1            AB-              Buck-Brown            donald76      571   
2             B+            Acosta-Black          diazrhonda     1368   
3             A+                加藤農林合同会社         nanamiinoue     1011   
4             A-  Ferragni-Rapisardi SPA        gaetanoboito     1662   
...          ...                     ...                 ...      ...   
2064          A-                株式会社田中印刷         shotasasaki     1064   
2065          O-                Siqueira   nogueiravalentina      326   
2066          A-          Borghese Group  baldassarepapafava     1519   
2067          B-           Griffin-Lopez          butlercory     1174   
2068          B-           Collins Group            sramirez     2054   

                   ssn  residence                 name  \
0          38671502902        NaN        Yuri Mendonça   
1          648-52-7110        NaN       Amanda Douglas   
2          472-71-8574        NaN          Jordan Shea   
3          156-44-9296        NaN                伊藤 修平   
4     CMMSYV42C16A763E        NaN  Sig. Atenulf Polani   
...                ...        ...                  ...   
2064       345-99-9952        NaN                鈴木 結衣   
2065       51603748920        NaN         Nicole Gomes   
2066  PLCDNN36E04G289W        NaN         Pina Salieri   
2067       136-94-5197        NaN    Mr. Curtis Wilson   
2068       033-10-3873        NaN        Crystal Smith   

                            job  \
0                           NaN   
1                           NaN   
2                           NaN   
3                        高等学校教員   
4                           NaN   
...                         ...   
2064                        NaN   
2065                        NaN   
2066  Occupational psychologist   
2067                        NaN   
2068                        NaN   

                                     current_location  station_ID  
0      (Decimal('75.5622165'), Decimal('140.611314'))         705  
1       (Decimal('56.355046'), Decimal('-31.498637'))         387  
2     (Decimal('77.9571885'), Decimal('-141.860092'))         593  
3         (Decimal('5.799277'), Decimal('58.160125'))         429  
4     (Decimal('78.8543295'), Decimal('-101.613684'))         388  
...                                               ...         ...  
2064   (Decimal('-72.9690365'), Decimal('92.739752'))         602  
2065    (Decimal('87.080457'), Decimal('129.153101'))         444  
2066   (Decimal('17.415553'), Decimal('-100.277465'))         287  
2067    (Decimal('52.450246'), Decimal('-96.122513'))         129  
2068   (Decimal('-40.630435'), Decimal('104.760671'))          61  

[2068 rows x 13 columns]

### 2. Analýza atribútov

In [123]:
print('Pocet atributov typu int a float:',  len(df_patient.select_dtypes(include=['int64','float64']).columns))

Pocet atributov typu int a float: 3


In [124]:
df_patient.select_dtypes(include=['int64','float64']).columns

Index(['user_id', 'residence', 'station_ID'], dtype='object')

In [125]:
df_patient.select_dtypes(include=['int64','float64']).nunique()

user_id       1315
residence        0
station_ID     767
dtype: int64

In [126]:
print('Pocet atributov typu object:', len(df_patient.select_dtypes(include=['object']).columns))

Pocet atributov typu object: 10


In [127]:
df_patient.select_dtypes(include=['object']).columns

Index(['mail', 'birthdate', 'registration', 'blood_group', 'company',
       'username', 'ssn', 'name', 'job', 'current_location'],
      dtype='object')

In [128]:
df_patient.select_dtypes(include=['object']).nunique()

mail                2062
birthdate           1119
registration        1875
blood_group            8
company             1885
username            2042
ssn                 2069
name                2018
job                  406
current_location    1966
dtype: int64

### 3. Vysledok analýzy
- dátumy ako `birth_date` a `registration` sú ako object --> mali by sa konvertovať na datetime
- `current_location` je ako object --> mali by sa rozložiť na dve číselné premenné (latitude, longitude) typu `float64`
- atribút `residence` má veľa hodnot typu NaN

## Dataset Station
### 1. Prieskum dát

In [129]:
df_station.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 831 entries, 0 to 830
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   code       829 non-null    object 
 1   QoS        831 non-null    object 
 2   longitude  831 non-null    float64
 3   revision   831 non-null    object 
 4   location   831 non-null    object 
 5   latitude   831 non-null    float64
dtypes: float64(2), object(4)
memory usage: 39.1+ KB


In [130]:
df_station.isnull().sum()

code         2
QoS          0
longitude    0
revision     0
location     0
latitude     0
dtype: int64

In [131]:
df_station.head()

code        QoS  longitude              revision           location  \
0   CN       good  117.33333  03/21/2024, 00:00:00      Asia/Shanghai   
1   IT       good   13.57661            2022/02/24        Europe/Rome   
2   US  excellent  -82.07842           16 Dec 2021   America/New_York   
3   BR       good  -35.71667           09 Feb 2020     America/Recife   
4   UA       good   36.25361            2019/01/30  Europe/Zaporozhye   

   latitude  
0  34.85000  
1  37.31065  
2  26.91756  
3  -8.66667  
4  47.48444

In [132]:
df_station.nunique()

code         112
QoS            4
longitude    558
revision     803
location     147
latitude     558
dtype: int64

In [133]:
df_station.describe()

longitude    latitude
count  831.000000  831.000000
mean    14.827578   28.049053
std     71.465292   23.426477
min   -123.026210  -54.800000
25%    -45.947220   13.542460
50%     13.576610   34.864720
75%     72.895155   45.766750
max    171.253640   68.798330

### 2. Analýza atribútov

In [134]:
df_station.select_dtypes(include=['float64']).columns

Index(['longitude', 'latitude'], dtype='object')

In [135]:
df_station.select_dtypes(include=['float64']).nunique()

longitude    558
latitude     558
dtype: int64

In [136]:
df_station.select_dtypes(include=['object']).columns

Index(['code', 'QoS', 'revision', 'location'], dtype='object')

In [137]:
df_station.select_dtypes(include=['object']).nunique()

code        112
QoS           4
revision    803
location    147
dtype: int64

### 3. Vysledok analýzy